### ISSUE TITLE
XGBoost eval_metric and prediction results are different

### ISSUE THREAD

**Issue by aParsecFromFuture** [dmlc/xgboost#8774]
```
def acc(y_true, y_pred):
        return accuracy_score(y_true, y_pred.round())

  params = {
      'objective': 'binary:logistic',
      'enable_categorical': True,
      'tree_method': 'hist',
      'eval_metric': acc
  }

  model = xgb.XGBClassifier(**params)

  model.fit(X_train, y_train, 
            eval_set=[(X_valid, y_valid)])
  
  y_pred = model.predict(X_valid)
  score = accuracy_score(y_valid, y_pred)
  print(score)
```

Accuracy score should be same score of the last iteration but I get 73% and 69%. Validation data has 10k rows. Any idea?

**Comment by trivialfis** [comment]
I can't reproduce using the latest xgboost. Could you please share a reproducible example?

**Comment by aParsecFromFuture** [comment]
I get the same behavior with built-in error metric. (1 - error) is not equal to accuracy score.

Class distribution: (0: 32496, 1: 20523)

<img width="678" alt="Screenshot 2023-02-14 at 22 31 37" src="https://user-images.githubusercontent.com/30957924/218838491-81a0ac93-1b23-4e84-99b6-d23c315ab2eb.png">

**Comment by trivialfis** [comment]
Let me get back to this tomorrow. I have an account with the same email address on my github profile.

**Comment by aParsecFromFuture** [comment]
> Let me get back to this tomorrow. I have an account with the same email address on my github profile.

I shared the notebook with trivialfis user. Good luck.

**Comment by trivialfis** [comment]
```
This is a limited-participation competition. Only invited users may participate.
```

**Comment by aParsecFromFuture** [comment]
> This is a limited-participation competition. Only invited users may participate.

I'm going to open it publicly for today. Let me know when your review is done.
Link: https://www.kaggle.com/greysky/temporary-notebook-it-will-be-deleted

**Comment by trivialfis** [comment]
thank you for sharing, I can see the notebook, it's the dataset that I can't access. I need to participate in the competition before seeing the dataset, but the competition is invite only.
![Screenshot from 2023-02-15 16-05-22](https://user-images.githubusercontent.com/16746409/218969104-1c188e92-039e-4002-aeb5-ab3c6187cbf0.png)

**Comment by aParsecFromFuture** [comment]
> thank you for sharing, I can see the notebook, it's the dataset that I can't access. I need to participate in the competition before seeing the dataset, but the competition is invite only.
> ![Screenshot from 2023-02-15 16-05-22](https://user-images.githubusercontent.com/16746409/218969104-1c188e92-039e-4002-aeb5-ab3c6187cbf0.png)

I have created the same notebook with copy-dataset and invited you. Could you try again?

**Comment by trivialfis** [comment]
Hi, I think it's possible that this is caused by the `verbose=20` in the `fit` method, the latest evaluation result is not displayed as the model print only at each 20 iterations.

You can obtain the list of evaluation results by `clf.evals_result()`.

**Comment by trivialfis** [comment]
NVM, later iteration is changed, still looking.

**Comment by aParsecFromFuture** [comment]
> Hi, I think it's possible that this is caused by the `verbose=20` in the `fit` method, the latest evaluation result is not displayed as the model print only at each 20 iterations.
> 
> You can obtain the list of evaluation results by `clf.evals_result()`.

There is no iteration with output accuracy. I noticed early that even if we do hyper parameter optimization (like Optuna), the output accuracy remains 0.61 as eval_metric output approaches 0.70. I was thinking that this happens because the LogLoss value is constantly increasing. But "disable_default_eval_metric" option didn't prevent that strange behavior.

**Comment by trivialfis** [comment]
Thank you for raising the issue and for the helpful assistance! I have reproduced the error using the dataset from the competition.
